pkg/streamingpromql/operator/instant_vector_selector.go

```diff
@@ -89,7 +85,23 @@ func (v *InstantVectorSelector) NextSeries(_ context.Context) (InstantVectorSeri
 			continue
 		}
 
-		data.Floats = append(data.Floats, promql.FPoint{T: stepT, F: val})
+		// if (val, h have been set by PeekPrev, we do not know if val is 0 because that's the actual value, or because
+		// the previous value had a histogram.
+		// PeekPrev will set the histogram to nil, or the value to 0 if the other type exists.
+		// So check if histograms is nil first. If we don't have a histogram, then we should have a value and vice-versa.
+		if h != nil {
+			if len(data.Histograms) == 0 {
+				// Only create the pool once we know the series is a histogram or not
```

Comment:
[nit]

```suggestion
				// Only create the slice once we know the series has histogram or not
```

(same applies below)

Reply:
It might also be worth mentioning that this will over-allocate in the case where we have both floats and histograms (we'll create a full-length slice for both in this case), but that's expected to be a rare case, so this is OK.